### PR TITLE
Added undefined check when accessing style.paddingLeft in dynamic-column-width implementation

### PR DESCRIPTION
**Pull request by t00**
Ngrid is crashing in the dynamic-column-width implementation when column width is not configured. Presumably style is not created and the logic expects a column with defined style.
The code is pretty complex and potentially it should not happen - maybe style not being set is an unexpected situation and setting the style should be fixed, yet surely crashing was not intended so I added checks for column style.

Potentially this is related to another issue - when column widths are defined in a way that one column has no width defined, and only minWidth set, when scrolling a longer lists columns get shifted.

Is it mandatory for all columns to have width set to percentage if any of the columns have percentage set? Image attached to show a problem.
image
image

### CODE REVIEW

**Comment by t00** [comment]
Regarding the varying column widths when scrolling a longer list - this is related to setting minWidth to some or all of the columns in addition to usual percentage split.
Is there a better way to set columns proportionally in ngrid in a way, so that minimum column width will be enforced and if browser window is not big enough, scroll viewer will show up and allow horizontal scrolling?

There is also another issue I have noticed - when minWidth is set only for one column (all have % width), grid width exceeds the container causing horizontal scrollbar to show. Maybe margins are somewhere not included in the calculation.

Also on mobile sometimes headers are lagging behind body when horizontal scrollbar is present.

As a side note - pbl-ngrid is great, using latest tech and very extendible so I am hoping to stay for longer and at least help to "unbug" it. Coming from ngx-datatable (slow) and mat-table (sticky header issues and a lack of easily implemented standard grid features). Thanks Shlomi.

**Comment by shlomiassaf** [comment]
@t00, again, thanks for taking the time, getting into the code and contributing, this is highly appreciated!

You've mentioned several issues or potential issue in this PR, mainly in the 2nd comment.

I can't address them all here, so i'll split them into issues and I'd love to hear additional information if you have such...

**Comment by shlomiassaf** [comment]
> Also on mobile sometimes headers are lagging behind body when horizontal scrollbar is present.

This is due to the fact the the headers in **fixed** mode are physically located outside of the grid data rows areas which is also the scroll area. So to make headers scroll when data cells horizontally scroll we listen to the event and act upon it. It works well on desktop, perhaps on mobile there's a different even we should listen to?

<details>
<summary>More details...</summary>
<p>
If we put header in the scroll area it will result in headers being part of vertical scrolling which means that they will disappear when scrolled out of the viewport. This is actually the <strong>row</strong> mode for headers which renders the header rows like regular data rows.
</p>
<p> 
Of course, you can also use the <strong>sticky</strong> mode which will render header cells within the scrolling area but will also apply sticky logic on them so they won't render outside of the viewport, but this does not work well with virtual scrolling and I really don't recommend it.
</p>
<p>
Note that this is a known limitation within the angular/cdk package which does not support the table's sticky feature along with the virtual scroll features (among with more unsupported things when you combine CDK components together)
</details>
   
See issue #56

**Comment by shlomiassaf** [comment]
> Potentially this is related to another issue - when column widths are defined in a way that one column has no width defined, and only minWidth set, when scrolling a longer lists columns get shifted.

Please see #57 , I couldn't reproduce this using the same column definitions (width wise)...

**Comment by shlomiassaf** [comment]
> There is also another issue I have noticed - when minWidth is set only for one column (all have % width), grid width exceeds the container causing horizontal scrollbar to show. Maybe margins are somewhere not included in the calculation.

Please see #58

**Comment by shlomiassaf** [comment]
As for the core issue, i'm aware of it.

This happens because of the `resize-observer` firing resize evert to DOM elements without a style, not even a style object that exists on the DOM element instance.

I suspect that it happens on DOM elements registered for changes but they also got destroyed yet to have their listener removed... or the other way around, created but still not initialized.

This has an affect on all areas, from fast scrolling to drag & drop and more...

So far, my approach to this was to fix the issue at the core cause this error is just a symptom...  
That is, to find why it fires and prevent it from firing at the source, and not here.

I tried to avoid this way of fixing because it will "hide" all other issues, firing events without need or missing de-registrations of elements...

I'm leaning towards accepting this solution because it does has a broad effect...

What do you think?

**Comment by shlomiassaf** [comment]
Also, about the width system, as time goes I keep thinking that i've made a mistake with how I've implemented percentage width's...

> Im writing this comment as a future preparation to better docs about width and starting of a design thinking session on how to tackle percentage width. This is for me but feel free to jump in!

This is pretty complex (especially when supporting both % and absolute px widths)...

---

Width is defined in the **width** property which accepts an absolute pixel value or relative percent value.

In addition, a numeric absolute (px) values of **minWidth** and/or **maxWidth** can be set, creating min/max width boundaries on top of the width.

The logic has some steps:

- Initial analysis of the columns, automatically completing blank widths.
- Calculating the proper width of the row container for proper horizontal scrolling.
- Handling resize events to columns and re-arranging width's accordingly.

There are 3 ways to define the **width**:

1. Leaving it blank
2. **px**, e.g. `50px`
3. **%**, e.g. `20%`

When left blank, the grid will calculate the width and use a relative width definition using the css `calc()` operator.

When fixed pixels are set, the width is fixed to that number

When a % is used, that value is set in CSS as is.

For example, 3 columns:

| Col        | A    | B   | C                |
|------------|------|-----|------------------|
| Definition | 50px | 25% | Blank            |
| Result     | 50px | 25% | calc(75% - 50px) |

More complex one:

| Col        | A    | B   | C                |     |                  |      |                  |                  |
|------------|------|-----|------------------|-----|------------------|------|------------------|------------------|
| Definition | 50px | 25% | Blank            | 35% | Blank            | 30px | Blank            | Blank            |
| Result     | 50px | 25% | calc(10% - 20px) | 35% | calc(10% - 20px) | 30px | calc(10% - 20px) | calc(10% - 20px) |

The logic for `calc()`:

- Sum all fixed **px** and fixed **%**
- Invert **%**, i.e. `100% - SUM%`
- Divide each result by the total number of blank column, let's prefix that result with **D**, so we have **Dpx** and **D%**
- Each cell will get that result as `calc(D% - Dpx)`

Our total px are `50 + 30 = 80` and total % are `25% + 35% = 60%`.  
Inverting 60% will get us to 40%.  
We have 4 blanks, so `80/4 = 20px` and `40/4 = 10%`.
Each blank will get `calc(10% - 20px)`

> This logic is implemented in the static width strategy: `StaticColumnWidthLogic`

Having the width's set is not enough, since all columns are contained within scroll containers we need to implement logic to properly handle horizontal scrolling.

To illustrate, in grid 500px wide, if the total **minWidth** of all columns is larger than 500px then we must force a horizontal scroll (done by setting min width of the internal scroll container).

> In a world with only absolute pixel width's that's not required, but we support minWidth on % width's

The static width strategy does this calculation as well.

---

This usually works fine but when using direct % widths there are some issues, especially when combining fixed pixel width assignments, minWidth's, etc...

The problem is that a relative column (%) does not play nicely with absolute width (px) since they are not on the same level (normalization is required).

This is a problem especially when horizontal scroll is required, and in such case a relative width from a calculated total scroll width will be hard to calculate cause each bump to the total scroll width will bump the actual width of the relative width column... hard logic...

In addition, it does not make sense to let the user define 15% and 15px and use them as is while blank will get only what's left (through `calc()`).  
In other words, by allowing `15%` and `15px` we allow fixating different worlds, we can only do one so I will probably only hard fix absolute widths (px) and convert all relative ones to `calc()` as well.

It means that setting 15% will mean 15% of the left over after reducing all absolute widths.  
E.G: Grid is 100px, col A is 20px, col B is 50%, col C is 50% --> A=20, B=(100-20)*50%=40, C=40.

In the current implementation, 20, 50% and 50% are on the same level so A=20, B=50, C=50 which is **larger** than 100!!!

In the new approach we are within the boundary.

**Comment by t00** [comment]
After refactoring I could not reproduce the original problem to which this PR was created. It is probably safe to close it.
I completely agree that if style is expected to be set, silently handling null check is not a way to go. Maybe logging error in a console is style is undefined would also be an option to track such issues without breaking the application if such issue was not found when testing? This will leave app in an undefined state with a log trace available and only a possibility that the issue is a showstopper for a user.

**Comment by t00** [comment]
Thank you for a thorough explanation of column width calculations, not sure if it is still valid after refactoring but looks like it does what it should, better yet using css when possible to scale the cells.

**Comment by t00** [comment]
Please feel free to close the PR as it might not be relevant now.
